packages/flutter/lib/src/widgets/dual_transition_builder.dart

```diff
@@ -0,0 +1,200 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+
+/// Builder callback used by [DualTransitionBuilder].
+///
+/// The builder is expected to return a transition powered by the provided
+/// `animation` and wrapping the provided `child`.
+///
+/// The `animation` provided to the builder always runs forward from 0.0 to 1.0.
+typedef AnimatedTransitionBuilder = Widget Function(
+  BuildContext context,
+  Animation<double> animation,
+  Widget child,
+);
+
+/// A transition builder that animates its [child] based on the
+/// [AnimationStatus] of the provided [animation].
+///
+/// This widget can be used to specify different enter and exit transitions for
+/// a [child].
+///
+/// While the [animation] runs forward, the [child] is animated according to
+/// [forwardBuilder] and while the [animation] is running in reverse, it is
+/// animated according to [reverseBuilder].
+///
+/// Using this builder allows the widget tree to maintain its shape by nesting
+/// the enter and exit transitions. This ensures that no state information of
+/// any descendant widget is lost when the transition starts or completes.
+class DualTransitionBuilder extends StatefulWidget {
```

Comment:
If we do bring the builder in here, we should bring in the tests from the animations package as well:
https://github.com/flutter/packages/blob/master/packages/animations/test/dual_transition_builder_test.dart

Reply:
I'll wait until we have conclusion from Goderbauer about the copy.

Reply:
+1